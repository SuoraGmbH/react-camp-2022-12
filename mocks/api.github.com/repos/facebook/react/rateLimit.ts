import { rest } from "msw";

export const githubRepoFacebookReactRateLimitMock = rest.get(
  "https://api.github.com/repos/facebook/react",
  (req, res, context) => {
    return res(
      context.status(403),
      context.delay(5000),
      context.json({
        message:
          "API rate limit exceeded for 93.245.109.13. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)",
        documentation_url:
          "https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting",
      })
    );
  }
);
