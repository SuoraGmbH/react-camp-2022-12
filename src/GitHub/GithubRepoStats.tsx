import React, { useEffect, useState } from "react";
import { GithubRepoData } from "./GithubRepoData";
import useGithubRepoStats from "./useGithubRepoStats";

interface Props {
  repoName: string;
}

const GithubRepoStats: React.FunctionComponent<Props> = ({ repoName }) => {
  const { repoData, error } = useGithubRepoStats(repoName);

  if (error) {
    return (
      <div>
        Error for repo {repoName}: {error}
      </div>
    );
  }

  if (!repoData) {
    return <div>Loadingâ€¦</div>;
  }

  return <div>facebook/react hat {repoData.stargazers_count} stars</div>;
};

export default GithubRepoStats;
