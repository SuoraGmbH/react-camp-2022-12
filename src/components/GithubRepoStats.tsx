import React, { useEffect, useState } from "react";
import { GithubRepoData } from "../domain/GithubRepoData";

const GithubRepoStats: React.FunctionComponent = () => {
  const [repoData, setRepoData] = useState<GithubRepoData>();
  const [error, setError] = useState<string>();

  useEffect(() => {
    fetch("https://api.github.com/repos/facebook/react").then(function (
      response
    ) {
      response.json().then(function (data: GithubRepoData) {
        if (response.status !== 200) {
          setRepoData(undefined);
          // I need to ignore a typescript error here,
          // because our typings do not contain the error case
          // @ts-expect-error
          setError(data?.message || "Something went wrong");
          return;
        }

        setError(undefined);
        setRepoData(data);
      });
    });
  }, []);

  if (error) {
    return <div>Error: {error}</div>;
  }

  if (!repoData) {
    return <div>Loadingâ€¦</div>;
  }

  return <div>facebook/react hat {repoData.stargazers_count} stars</div>;
};

export default GithubRepoStats;
