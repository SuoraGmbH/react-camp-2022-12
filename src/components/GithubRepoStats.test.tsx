import { render, screen } from "@testing-library/react";
import GithubRepoStats from "./GithubRepoStats";
import { setupServer } from "msw/node";
import { githubRepoFacebookReactSuccessMock } from "../../mocks/api.github.com/repos/facebook/react/success";
import { githubRepoFacebookReactRateLimitMock } from "../../mocks/api.github.com/repos/facebook/react/rateLimit";

const mockServer = setupServer();
beforeAll(() => {
  mockServer.listen({ onUnhandledRequest: "error" });
});

beforeEach(() => {
  mockServer.resetHandlers();
});

afterAll(() => {
  mockServer.close();
});

describe("<GithubRepoStats />", () => {
  // test("Smoke Test", () => {
  //   render(<GithubRepoStats />);
  // });
  //
  // test("Initially shows the loading indicator", () => {
  //   render(<GithubRepoStats />);
  //
  //   expect(screen.getByText(/loading/i)).toBeInTheDocument();
  // });

  test("The loading indicator should disappear", async () => {
    mockServer.resetHandlers(githubRepoFacebookReactSuccessMock);
    render(<GithubRepoStats />);

    await screen.findByText(/facebook\/react/i);
  });
  test("The number of stargazers should be visible", async () => {
    mockServer.resetHandlers(githubRepoFacebookReactSuccessMock);
    const { container } = render(<GithubRepoStats />);

    // Wait for loading indicator to disappear
    await screen.findByText(/facebook\/react/i);

    expect(container).toMatchSnapshot();
  });

  test("Display an error if we are rate limitted", async () => {
    mockServer.resetHandlers(githubRepoFacebookReactRateLimitMock);
    const { container } = render(<GithubRepoStats />);

    expect(container).toMatchSnapshot();
  });
});
