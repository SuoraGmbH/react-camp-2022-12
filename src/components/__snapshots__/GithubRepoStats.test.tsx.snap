// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GithubRepoStats /> Display an error if we are rate limitted 1`] = `
<div>
  <div>
    Loadingâ€¦
  </div>
</div>
`;

exports[`<GithubRepoStats /> The number of stargazers should be visible 1`] = `
<div>
  <div>
    facebook/react hat 
    199045
     stars
  </div>
</div>
`;
